### PR TITLE
driver management implementation

### PR DESCRIPTION
**Pull request by tonistiigi**
Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

```
/work # docker buildx

Usage:	docker buildx COMMAND

Build with BuildKit

Commands:
  bake        Build from a file
  build       Start a build
  create      Create a new builder instance
  inspect     Inspect current builder instance
  ls          List builder instances
  rm          Remove a builder instance
  use         Set the current builder instance

Run 'docker buildx COMMAND --help' for more information on a command.
/work # docker buildx create --help

Usage:	docker buildx create [OPTIONS] [CONTEXT|ENDPOINT]

Create a new builder instance

Options:
      --append                 Append a node to builder instead of changing it
      --driver string          Driver to use (eg. docker-container)
      --leave                  Remove a node from builder instead of changing it
      --name string            Builder instance name
      --node string            Create/modify node with given name
      --platform stringArray   Fixed platforms for current node
/work # docker buildx use --help

Usage:	docker buildx use [OPTIONS] NAME

Set the current builder instance

Options:
      --default   Set builder as default for current context
      --global    Builder persists context changes
/work # docker buildx inspect --help

Usage:	docker buildx inspect [NAME]

Inspect current builder instance

Options:
      --bootstrap   Ensure builder as booted before inspecting

/work # docker context create my-context  --docker "host=unix:///var/run/docker.sock"

my-context
Successfully created context "my-context"
/work #
/work # docker buildx create --use my-context
gifted_kapitsa
/work # docker buildx inspect --bootstrap
[+] Building 12.0s (1/1) FINISHED
 => [internal] booting buildkit                                                                     11.9s
 => => pulling image moby/buildkit:master                                                           10.8s
 => => creating container buildx_buildkit_gifted_kapitsa0                                            1.2s
Name:   gifted_kapitsa
Driver: docker-container

Nodes:
Name:      gifted_kapitsa0
Endpoint:  my-context
Status:    running
Platforms: linux/amd64, linux/arm64, linux/arm/v7, linux/arm/v6
/work #
/work # docker buildx ls
NAME              DRIVER/ENDPOINT  STATUS  PLATFORMS
gifted_kapitsa *  docker-container
  gifted_kapitsa0 my-context       running linux/amd64, linux/arm64, linux/arm/v7, linux/arm/v6
my-context        docker
  my-context      my-context       running linux/amd64, linux/arm64, linux/arm/v7, linux/arm/v6
default           docker
  default         default          running linux/amd64, linux/arm64, linux/arm/v7, linux/arm/v6
```

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	root string
+}
+
+func (s *Store) Txn() (*Txn, func(), error) {
+	l := flock.New(filepath.Join(s.root, ".lock"))
+	if err := l.Lock(); err != nil {
+		return nil, nil, err
+	}
+	return &Txn{
+			s: s,
```
Comment (tiborvass) [inline store/store.go]: is this correct gofmt ?

Comment (tonistiigi) [reply store/store.go]: it does not error

File: commands/use.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+				}
+				return nil
+			}
+			list, err := dockerCli.ContextStore().ListContexts()
+			if err != nil {
+				return err
+			}
+			for _, l := range list {
+				if l.Name == name {
+					return errors.Errorf("to switch to context %s use `docker context use %s`", name, name)
```
Comment (tiborvass) [inline commands/use.go]:
```
Run `docker context use %s` to switch to context %s
```

File: store/nodegroup.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		return errors.Errorf("invalid duplicate endpoint %s", ep)
+	}
+	return nil
+}
+
+func (ng *NodeGroup) findNode(name string) int {
+	i := -1
+	for ii, n := range ng.Nodes {
+		if n.Name == name {
+			i = ii
```
Comment (tiborvass) [inline store/nodegroup.go]: no early return?

File: commands/create.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	actionLeave  bool
+	use          bool
+	// upgrade      bool // perform upgrade of the driver
+}
+
+func runCreate(dockerCli command.Cli, in createOptions, args []string) error {
+	ctx := appcontext.Context()
+
+	if in.name == "default" {
+		return errors.Errorf("default is a reserved name and can't be used to identify builder instance")
```
Comment (tiborvass) [inline commands/create.go]: `can't` -> `cannot`

File: commands/inspect.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	err = loadNodeGroupData(timeoutCtx, dockerCli, ngi)
+
+	if in.bootstrap {
+		if err := boot(ctx, ngi); err != nil {
+			return err
+		}
+		ngi = &nginfo{ng: ng}
```
Comment (tiborvass) [inline commands/inspect.go]: dupe

Comment (tonistiigi) [reply commands/inspect.go]: no, we need to reload from clean state after boot

**Comment by tiborvass** [comment]
Not fond of the assymmetry between `--append` and `--leave` (`--add`/`--rm` ?), I don't like `--node` because it has too little to do with `docker node` (how about `worker`? or `agent`?), and I think `view` should eventually go away from this command.

My understanding after going through UI with Tonis is that 99% of the usecases should be covered with `create`, `use` and `--append`, the rest is advanced usage and extremely rare, which excuses the overloading of the `create` command to save confusion from having too many management commands for very rare usecases.

**Comment by tiborvass** [comment]
Ping @dmcgowan

**Comment by tonistiigi** [comment]
> (--add/--rm ?)

add has much vaguer meaning and it isn't clear that resources are being inserted into an existing set (could very well mean that just another builder instance is added to ls). rm can be confused with `buildx rm` that is a different use-case.

This used to be `--join/--leave` but I updated it after design feedback.

**Comment by tonistiigi** [comment]
As @tiborvass said, If there are more design comments leave them here or open new issues. We're in a phase where we can still revert things.

**Pull request merged**